servicetalk-http-api/src/main/java/io/servicetalk/http/api/AbstractZipContentCodec.java

```diff
@@ -0,0 +1,540 @@
+package io.servicetalk.http.api;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.Publisher;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.zip.CRC32;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterOutputStream;
+import java.util.zip.Inflater;
+import java.util.zip.InflaterInputStream;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.buffer.api.Buffer.asInputStream;
+import static io.servicetalk.buffer.api.Buffer.asOutputStream;
+import static io.servicetalk.buffer.api.EmptyBuffer.EMPTY_BUFFER;
+import static io.servicetalk.concurrent.api.Single.succeeded;
+import static io.servicetalk.http.api.HeaderUtils.addContentEncoding;
+import static java.lang.Math.min;
+
+abstract class AbstractZipContentCodec implements ContentCodec {
+
+    protected static final int ONE_KB = 1 << 10;
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractZipContentCodec.class);
+    private static final Object END_OF_STREAM = new Object();
+
+    private final CharSequence encoding;
+
+    AbstractZipContentCodec(final CharSequence encoding) {
+        this.encoding = encoding;
+    }
+
+    abstract boolean supportsChecksum();
+
+    abstract Inflater newRawInflater();
+
+    abstract DeflaterOutputStream newDeflaterOutputStream(OutputStream out) throws IOException;
+
+    abstract InflaterInputStream newInflaterInputStream(InputStream in);
+
+    @Override
+    public final Buffer encode(final HttpHeaders headers, final Buffer src, final int offset, final int length,
+                               final BufferAllocator allocator) {
+        addContentEncoding(headers, encoding);
+
+        final Buffer dst = allocator.newBuffer(ONE_KB);
+        DeflaterOutputStream output = null;
+        try {
+            output = newDeflaterOutputStream(asOutputStream(dst));
+
+            if (src.hasArray()) {
+                output.write(src.array(), offset, length);
+            } else {
+                while (src.readableBytes() > 0) {
+                    byte[] onHeap = new byte[min(src.readableBytes(), ONE_KB)];
+                    src.readBytes(onHeap);
+                    output.write(onHeap);
+                }
+            }
+
+            output.finish();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } finally {
+            closeQuietly(output);
+        }
+
+        return dst;
+    }
+
+    @Override
+    public final Publisher<Buffer> encode(final HttpHeaders headers, final Publisher<Buffer> from,
+                                          final BufferAllocator allocator) {
+        addContentEncoding(headers, encoding);
+        return from
```

Comment:
Hold on reviewing this part, after some discussions, I need to consider controlling demand from within. I am looking at the `Processor` API to see how this can be hooked.